tests/core/test_funcs.py

```diff
@@ -0,0 +1,71 @@
+import uuid
+from typing import Optional
+
+import pytest
+
+from dls_bluesky_core.core import group_uuid, in_micros, step_to_num
+
+
+@pytest.mark.parametrize(
+    "s,us",
+    [
+        (4.000_001, 4_000_001),
+        (4.999_999, 4_999_999),
+        (4, 4_000_000),
+        (4.000_000_1, 4_000_001),
+        (4.999_999_9, 5_000_000),
+        (0.1, 100_000),
+        (0.000_000_1, 1),
+        (0, 0),
+    ],
+)
+def test_in_micros(s: float, us: int):
+    assert in_micros(s) == us
+
+
+@pytest.mark.parametrize(
+    "s", [-4.000_001, -4.999_999, -4, -4.000_000_5, -4.999_999_9, -4.05]
+)
+def test_in_micros_negative(s: float):
+    with pytest.raises(ValueError):
+        in_micros(s)
+
+
+@pytest.mark.parametrize(
+    "start,stop,step,expected_num,truncated_stop",
+    [
+        (0, 0, 1, 1, None),  # start=stop, 1 point at start
+        (0, 0.5, 1, 1, 0),  # step>length, 1 point at start
+        (0, 1, 1, 2, None),  # stop=start+step, point at start & stop
+        (0, 0.99, 1, 2, 1),  # stop >= start + 0.99*step, included
+        (0, 0.98, 1, 1, 0),  # stop < start + 0.99*step, not included
+        (0, 1.01, 1, 2, 1),  # stop >= start + 0.99*step, included
+        (0, 1.75, 0.25, 8, 1.75),
+        (0, 0, -1, 1, None),  # start=stop, 1 point at start
+        (0, 0.5, -1, 1, 0),  # abs(step)>length, 1 point at start
+        (0, -1, 1, 2, None),  # stop=start+-abs(step), point at start & stop
+        (0, -0.99, 1, 2, -1),  # stop >= start + 0.99*-abs(step), included
+        (0, -0.98, 1, 1, 0),  # stop < start + 0.99*-abs(step), not included
+        (0, -1.01, 1, 2, -1),  # stop >= start + 0.99*-abs(step), included
+        (0, -1.75, 0.25, 8, -1.75),
+    ],
```

Comment:
(1, 10, -0.901, 10, 0.99),        This will fail currently and actually give (1, 10, 9, 1.891)
(10, 1, -0.901, 10, 0.99),        This will pass

The only negative step cases tested above are when start and stop are equal or when the step is bigger than abs(stop - start), there should be some true negative step cases added

Reply:
Added test case with the agreed upon behaviour, to ensure it doesn't change